lib/src/entrypoint.dart

```diff
@@ -62,21 +63,78 @@ class Entrypoint {
   // true.
   final String workingDir;
 
-  Package? _workspaceRoot;
+  static ({Package root, Package work}) _loadWorkspace(
+    String workingDir,
+    SystemCache cache,
+  ) {
+    if (!dirExists(workingDir)) {
+      fail('The directory `$workingDir` does not exist.');
+    }
+    // Keep track of all the pubspecs met when walking up the file system.
+    // The first of these is the workingPackage.
+    final pubspecsMet = <String, Pubspec>{};
+    for (final dir in parentDirs(workingDir)) {
+      final Pubspec pubspec;
+
+      try {
+        pubspec = Pubspec.load(
+          dir,
+          cache.sources,
+          containingDescription: RootDescription(dir),
+          allowOverridesFile: true,
+        );
+      } on FileException {
+        continue;
+      }
+      pubspecsMet[p.canonicalize(dir)] = pubspec;
+      final Package root;
+      if (pubspec.resolution == Resolution.none) {
+        root = Package.load(
+          null,
+          dir,
+          cache.sources,
+          alreadyLoadedPubspecs: pubspecsMet,
+          withPubspecOverrides: true,
+        );
+        for (final package in root.transitiveWorkspace) {
+          pubspecsMet.entries.first;
+          if (identical(pubspecsMet.entries.first.value, package.pubspec)) {
+            return (root: root, work: package);
+          }
+        }
+        assert(false);
+      }
+    }
+    if (pubspecsMet.isEmpty) {
+      throw FileException(
+        'Found no `pubspec.yaml` file in `${p.canonicalize(workingDir)}` or parent directories',
+        p.join(workingDir, 'pubspec.yaml'),
+      );
+    } else {
+      final firstEntry = pubspecsMet.entries.first;
+      throw FileException(
+        '''
+Found a pubspec.yaml at ${firstEntry.key}. But it has resolution `${firstEntry.value.resolution.name}`.
+But found no workspace root including it in parent directories.
+
+See $workspacesDocUrl for more information.''',
+        p.join(workingDir, 'pubspec.yaml'),
+      );
+    }
+  }
+
+  ({Package root, Package work})? _packages;
+  ({Package root, Package work}) get _getPackages =>
+      _packages ??= _loadWorkspace(workingDir, cache);
```

Comment:
Documenting these would be nice.

Specifically, documenting that: `_getPackages` returns, and that `_packages` should never be accessed, it's only for caching `_getPackages`, right?

Just a line for each explaining the relationship, and what `_getPackages` is supposed to be.

Reply:
Done